### ISSUE TITLE
variables not removed from the inventory when removed from source

### ISSUE THREAD

**Issue by cs-laurentiuvasiescu** [ansible/awx#5682]
##### ISSUE TYPE

 - Bug Report

##### SUMMARY

After removing variables from the original SCM project they will still be present in the inventory variables.

##### ENVIRONMENT

* AWX version: 9.1.0
* AWX install method: rpm
* Ansible version:  2.9.2
* Operating System: Centos 7,7
* Web Browser: Chrome

##### STEPS TO REPRODUCE

Create an SCM project, create an inventory sourced in that project (update options for the source: overwrite, overwrite variables, update on project update are checked).

Define few variables in group_vars/all.yml file.
update the project, sync the inventory. Variable is in the Details pane, under Variables. 

Delete one variable from group_vars/all.yml file in the SCM.
Update the project, sync the inventory. 

##### EXPECTED RESULTS

Variable is not defined anymore. Should not be in Details pane, under Variables.
Any play run against the playbook should not have access to the variable.

##### ACTUAL RESULTS

Variable is still there in Details pane, under Variables.
Any play run against the playbook still has access to the variable.

**Comment by wenottingham** [comment]
Is this variable at the inventory level? Vars at the toplevel inventory are not automatically cleared, as they can also be set manually outside of any one (or multiple) inventory sources.

**Comment by wenottingham** [comment]
Closing, please reopen if that's not the case.

**Comment by testArtest** [comment]
so, right now i have:
inventories
  - test
      - inventory.yml
      - group_vars
          - all.yml

And the variables defined in group_vars/all DO NOT get refreshed when i change them in the SCM. is there anyway to force them to update when i update the project they are sourced from?

I even deleted the variables on the details page of that inventory in tower and they didn't get repopulated after re-sync. 

I would expect the variables to be repopulated with variables from the project/SCM ...

**Comment by jibbajabber** [comment]
@wenottingham I'm also seeing this issue, just want to clarify if this is the intent?
My usecase:
* inventory source overwrite variables checked
* project update always for each job template run
* set some host var in sourced from project static inventory, say `my_remote: "some_remote_value"` in SCM
* run job template
* remove `my_remote`
* run job template again

Behaviour
* `my_remote` still present with original value after removal
* `my_remote` still present with original value after update

The host var value doesn't appear to change beyond the value first set in AWX, I can see project and inventory source update happening before job template is run. Even with manual sync of inventory source then job template run, symptom persists.

My inventory structure looks like this:
```
inventory/
  site1/
    some_host_invetory_file.yml
    some_other_host_inventory_file.yml
  site2/
  site3/
  group_vars/
  host_vars/
     host1.yml
     host2.yml
```
All host vars are set within the host_vars directory, similarly for group vars (I've not yet tested if group vars suffer from same symptom.

Could you clarify if this behaviour is the intent? 

AWX docker deployment using `13.0.0` tag

**Comment by vin-c** [comment]
Major issue for us if we need to manually clear all the variables at the inventory level as in our use-case, no var is setup here by anyone, all vars are coming from a merge-script inside our repo

/reopen

**Comment by ianlshaw** [comment]
I've run into this too.

If this is by design, it would be interesting to hear the rationale.

When using source based inventories, and certainly when checking "Overwrite" and "Overwrite Variables" - which seems like the correct method when relying on SCM based inventories, it seems counter intuitive that variables which are subsequently removed from the codebase aren't removed from the inventory.

The wording of the tooltip for "Overwrite Variables" is as follows:
"If checked, all variables for child groups and hosts will be removed and replaced by those found on the external source."

I would argue (perhaps incorrectly) that this _implies_ that a variable which once existed on the external source but no longer does "will be removed".  Which is not the behavior we see.

Pedantic wording aside. The issue this presents is one where:
If you rely heavily on external inventories, you cannot manage these inventories _exclusively_ in the codebase, and you must know that if a variable is removed you'll have to go through and manually remove it from each of your inventories before the single-source-of-truth(the repository) is restored.

As you can imagine, if you're dealing with hundreds of inventories, this could be a painstaking and human-error-prone procedure.

**Comment by kuznetsovkk** [comment]
Still relevant for me (awx 17.1.0). Openned new issue https://github.com/ansible/awx/issues/10200

**Comment by nlvw** [comment]
@wenottingham This issue should not be closed as it represents a major problem.   By not removing variables, when they are removed by the source, AWX is incorrectly running project code.

**Comment by BugBuster1701** [comment]
@wenottingham  We have the problem in AWX 15.0.1.  (variable ansible_become in group_vars)
Just lost 2 hours of time because of this bug.
New variables are added, so why are deleted ones not removed?
I don't understand your answer [here](https://github.com/ansible/awx/issues/5682#issuecomment-575735510)

**Comment by Tobias-Pe** [comment]
I see that this is not a Bug but an intended behavior because of manually added variables in awx.
In #13045 i suggest an opt-in to clear all variables before a sync, so us who disturbes this behavior have the option to always have mirrored variables in our inventories.

**Comment by jrgoldfinemiddleton** [comment]
Based on [this comment](https://github.com/ansible/awx/issues/13045#issuecomment-1501839510) by @AlanCoding it appears this bug is not resolved and is considered to be a bug still.

Could we please re-open this ticket so it gets addressed?
